### PR TITLE
Improve the handling of SVE state as part of threadsuspend

### PR DESCRIPTION
**Pull request by tannergooding**
This does not validate that all the handling in `pal/context.cpp` is correct. There's some known places it isn't handling full width registers and where it may be inconsistent with the Win32 CONTEXT layout.

### CODE REVIEW

**Comment by kunalspathak** [comment]
@JasonLinMS

File: src/coreclr/pal/src/arch/arm64/asmconstants.h
Comment (tannergooding) [reply src/coreclr/pal/src/arch/arm64/asmconstants.h:24]: The general `CONTEXT` struct as is currently used by PAL is supposed to mirror the win32 layout/defines so they can translate 1-to-1

The win32 naming convention here is then `XSTATE_ARM64_SVE` and it is using bit 2, not bit 1

File: src/coreclr/pal/inc/pal.h
Comment (tannergooding) [inline src/coreclr/pal/inc/pal.h:1849]: I think there might need to be an additional set of changes in `vm/arm64/asmconstants.h` as there is currently a `SIZEOF__CONTEXT` struct that exists and which is meant to represent the size of the struct on Win32 vs UNIX and I believe it is currently incorrect for Unix (or may need additional handling to ensure its large enough to hold the additional SVE state based on the vector length).

Comment (tannergooding) [reply src/coreclr/pal/inc/pal.h:1849]: That shouldn't be related to the Windows handling being added, however, so I've left it to a future PR

File: src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp
Comment (tannergooding) [inline src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:515]: This isn't defined for ARM64 until v10.0.26100.0; going to need to dynamically resolve it instead like we do for `pfnInitializeContext2`

Not going to do that tonight, just leaving a note for myself for tomorrow

File: src/coreclr/pal/src/arch/arm64/asmconstants.h
Comment (kunalspathak) [reply src/coreclr/pal/src/arch/arm64/asmconstants.h:24]: > The win32 naming convention here is then XSTATE_ARM64_SVE and it is using bit 2, not bit 1

So wrong bit was set for linux in https://github.com/dotnet/runtime/pull/103801 as well?

Comment (tannergooding) [reply src/coreclr/pal/src/arch/arm64/asmconstants.h:24]: Right.

I don't think it would technically cause problems because we're just emulating the CONTEXT struct and it doesn't actually have to match the Win32 struct 1-to-1 (since its only going to be other pal APIs consuming it)

But it's overall better/simpler to ensure they match so we don't end up with potential conflicts or other issues.

The best option here would be to just consume the native `context_t` struct from Unix directly (and not do any of this shimming), but that's a more involved PR.

Comment (jkotas) [reply src/coreclr/pal/src/arch/arm64/asmconstants.h:24]: > it doesn't actually have to match the Win32 struct 1-to-1

In fact, it does not match the Win32 struct 1-to-1: https://github.com/dotnet/runtime/pull/103801#discussion_r1649435005

> The best option here would be to just consume the native context_t struct from Unix directly (and not do any of this shimming), but that's a more involved PR.

+1. The Windows CONTEXT structure is very entrenched throughout CoreCLR VM.

File: src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp
Comment (janvorli) [reply src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp]: Do you refer to Windows SDK version or a Windows version?

Comment (tannergooding) [reply src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp]: Both, it's a new API for Arm64 in v10.0.261000.

Hence we try to resolve the export and don't do anything on Arm64 if it can't be resolved.

File: src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp
Comment (tannergooding) [inline src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp:482]: Windows doesn't define `CONTEXT_XSTATE` for arm64, it only defines `CONTEXT_ARM64_XSTATE`. So this is primarily to avoid a lot of other ifdefs elsewhere in the file.

**Comment by JulieLeeMSFT** [comment]
@MichalStrehovsky, could you review this PR before the preview 7 snap, 7/22 4pm PST?

**Comment by JulieLeeMSFT** [comment]
cc @kunalspathak for a code review.

File: src/coreclr/pal/inc/pal.h
Comment (janvorli) [reply src/coreclr/pal/inc/pal.h:1850]: I am not sure how it could be incorrect when there is an assert that it matches the sizeof (T_CONTEXT) in the asmconstants.h.

Comment (tannergooding) [reply src/coreclr/pal/inc/pal.h:1850]: The consideration is that SVE registers don’t have a fixed width and so the current logic asserting the size statically is already wrong (it’s assuming SVE registers are exactly 128-bits, when they can be up to 2048 bits)

Comment (janvorli) [reply src/coreclr/pal/inc/pal.h:1850]: That's true, but the constant is to represent the size of the CONTEXT structure and that's fixed for now, so it is correct. I believe that when we'll be adding support for longer SVE registers, we should finally move to the Windows way of separate context and extended context, so the SIZEOF__CONTEXT would still be constant - and it will actually match Windows one after that.

**Comment by janvorli** [comment]
A nit - the title of this PR says it is ensuring saving SVE state on Windows, but it also modifies quite a number of Unix only files. It would be nice to rename it.

**Review by janvorli — APPROVED** [review]
LGTM

**Review by TIHan — APPROVED** [review]

**Comment by tannergooding** [comment]
I believe this is just waiting on review from @MichalStrehovsky and otherwise should be good to merge

**Comment by kunalspathak** [comment]
> I believe this is just waiting on review from @MichalStrehovsky and otherwise should be good to merge

Not sure if @MichalStrehovsky might be around until tomorrow. @JulieLeeMSFT - since we already got 2 signoffs, can we merge this in?

**Comment by JulieLeeMSFT** [comment]
Yes, let's merge in.

**Comment by tannergooding** [comment]
This missed the snap and will need to be backported, so it will need to go through tactics for approval.

**Comment by JulieLeeMSFT** [comment]
Missed by a minute or so. Please go ahead with a tactics approval.

**Comment by MichalStrehovsky** [comment]
> @MichalStrehovsky, could you review this PR before the preview 7 snap, 7/22 4pm PST?

I know next to nothing about thread suspension, @VSadov FYI

**Pull request merged**
